conda/gateways/repodata/__init__.py

```diff
@@ -289,3 +290,96 @@ def conda_http_errors(url, repodata_fn):
             e.response,
             caused_by=e,
         )
+
+
+class CacheJsonState(dict):
+    """
+    Load/save `.state.json` that accompanies cached `repodata.json`
+    """
+
+    def __init__(self, cache_path_json, cache_path_state, repodata_fn):
+        self.cache_path_json = pathlib.Path(cache_path_json)
+        self.cache_path_state = pathlib.Path(cache_path_state)
+        self.repodata_fn = repodata_fn
+
+    def load(self):
+        """
+        Cache headers and additional data needed to keep track of the cache are
+        stored separately, instead of the previous "added to repodata.json"
+        arrangement.
+        """
+        try:
+            state_path = self.cache_path_state
+            log.debug("Load %s cache from %s", self.repodata_fn, state_path)
+            state = json.loads(state_path.read_text())
+            # json and state files should match
+            json_stat = self.cache_path_json.stat()
+            if not (
+                state.get("mtime_ns") == json_stat.st_mtime_ns
+                and state.get("size") == json_stat.st_size
+            ):
+                # clear mod, etag, cache_control to encourage re-download
+                state.update({"etag": "", "mod": "", "cache_control": "", "size": 0})
+            for field in (
+                "etag",
+                "mod",
+                "cache_control",
+                "size",
+            ):
+                alias = field
+                if alias in state:
+                    self[field] = state[alias]
+        except (json.JSONDecodeError, OSError):
+            log.debug("Could not load state", exc_info=True)
+            self.clear()
+        return self
+
+    def save(self):
+        """
+        Must be called after writing cache_path_json, as its mtime is included in .state.json
+        """
+        json_stat = self.cache_path_json.stat()
+        serialized = {"mtime_ns": json_stat.st_mtime_ns, "size": json_stat.st_size}
+        for field in (
+            "etag",
+            "mod",
+            "cache_control",
+            "size",
+        ):
+            alias = field
+            if field in self:
+                serialized[alias] = self[field]
+        return pathlib.Path(self.cache_path_state).write_text(json.dumps(serialized, indent=True))
+
+    @property
+    def mod(self) -> str:
+        """
+        Last-Modified header or ""
+        """
+        return self.get("mod", "")
+
+    @mod.setter
+    def mod(self, value):
+        self["mod"] = value or ""
```

Comment:
Per the way we serialize in `save` shouldn't these properties clear empty values?

```suggestion
        if value:
            self["mod"] = value
        elif "mod" in self:
            del self["mod"]
```